astrocut/cube_cut_from_footprint.py

```diff
@@ -0,0 +1,310 @@
+
+import json
+import re
+import os
+from typing import List, Literal, Union
+import warnings
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+from astropy.coordinates import SkyCoord
+from astropy.table import Table, Column
+import astropy.units as u
+from astropy.utils.exceptions import AstropyWarning
+import numpy as np
+import fsspec
+from spherical_geometry.polygon import SphericalPolygon
+from spherical_geometry.vector import radec_to_vector
+
+from astrocut.exceptions import InvalidQueryError
+from astrocut.cube_cut import CutoutFactory
+
+from .utils.utils import parse_size_input
+
+TESS_ARCSEC_PER_PX = 21
+
+
+def _s_region_to_polygon(s_region: Column):
+    """
+    Takes in a s_region string of type POLYGON or CIRCLE and returns it as
+    a spherical_region Polygon.
+
+    Example inputs:
+    'POLYGON 229.80771900 -75.17048500 241.67788000 -63.95992300 269.94872000 -64.39276400 277.87862300 -75.57754400'
+    'CIRCLE ICRS 244.38324081 -75.86611807 0.625'
+    """
+
+    def ind_sregion_to_polygon(s_reg):
+        sr_list = s_reg.strip().split()
+        reg_type = sr_list[0].upper()
+
+        if reg_type == "POLYGON":
+            ras = np.array(sr_list[1::2], dtype=float)
+            ras[ras < 0] = ras[ras < 0] + 360
+            decs = np.array(sr_list[2::2], dtype=float)
+            return SphericalPolygon.from_radec(ras, decs)
+        elif reg_type == "CIRCLE":
+            ra, dec, rad = np.array(sr_list[-3:], dtype=float)
+            return SphericalPolygon.from_cone(ra, dec, rad)
+        else:
+            raise ValueError("Unsupported S_Region type.")
+
+    return np.vectorize(ind_sregion_to_polygon)(s_region)
+
+
+def _get_s3_ffis(s3_uri, as_table: bool = False, load_polys: bool = False) -> Table | dict:
+    """
+    Fetch the S3 footprint file containing a dict of all FFIs and a polygon column
+    that holds the s_regions as polygon points and vectors.
+
+    Optional Parameters:
+        as_table: Return the footprint file as an Astropy Table
+        load_polys: Convert the s_region column to an array of SphericalPolygon objects
+    """
+    # Open footprint file with fsspec
+    # Use caching to help performance, but check that remote UID matches the local
+    # Expiry time is 1 week by default
+    with fsspec.open('filecache::' + s3_uri, s3={'anon': True}, filecache={'cache_storage':'s3_cache', 'check_files': True}) as f:
+        ffis = json.loads(f.read())
+
+    if load_polys:
+        ffis['polygon'] = _s_region_to_polygon(ffis['s_region'])
+
+    if as_table:
+        ffis = Table(ffis)
+
+    return ffis
+
+
+def _ffi_intersect(ffi_list: Table, polygon: SphericalPolygon):
+    """
+    Vectorizing the spherical_coordinate intersects_polygon function
+    """
+    def single_intersect(ffi, polygon):
+        return ffi.intersects_poly(polygon)
+
+    return np.vectorize(single_intersect)(ffi_list["polygon"], polygon)
```

Comment:
I tried added multiprocessing to this function, but found that there was no performance improvement (if anything, execution time went up because of thread overhead.) In any case, the `ra_dec_crossmatch` function takes < 1 sec to run.

Reply:
when you say threads, do you actually mean threads or processes?

This is CPU bound, and currently, with the GIL, there will not be any improvement using threads. You would need to use multiple processes.

Reply:
I did mean threads, but I just tried with `ProcessPoolExecutor`, and oddly enough, timing is a bit worse. It takes around 1 second longer with the process pool vs. vectorize. This is the code I'm using for comparison:

```
def single_intersect(ffi, polygon):
        return ffi.intersects_poly(polygon)


def _ffi_intersect_process(ffi_list: Table, polygon: SphericalPolygon, workers=None):
    """
    Vectorizing the spherical_coordinate intersects_polygon function
    """
    
    with ProcessPoolExecutor(max_workers=workers) as executor:
        # Map the function across the ffi_list["polygon"]
        results = list(executor.map(single_intersect, ffi_list["polygon"], [polygon] * len(ffi_list)))

    return np.array(results)
```

Reply:
I tried your attached code and couldn't get it to work in a different file without some modifications. I also am not sure it's correct, with the arguments you've specified in `map`.

However, I can confirm that using the process pool executor is no faster, and a tiny bit worse, no matter what number of processes I use. There is overhead to starting up subprocesses in Python, so perhaps that's the issue we're running into. Also, SphericalGeometry might have some multiprocessing itself, and so we're just getting in the way of that by starting multiple processes?

Running the below code reports the intersect taking just over 1 second on my machine, whereas the version in this PR takes around a second or just under a second.

```py
import functools
import sys
import time
from concurrent.futures import ProcessPoolExecutor
from multiprocessing import freeze_support

import numpy as np
from astropy.coordinates import SkyCoord
from astropy.table import Table
from spherical_geometry.polygon import SphericalPolygon

import astrocut.cube_cut_from_footprint

ccff = sys.modules["astrocut.cube_cut_from_footprint"]


def single_intersect(polygon, ffi_poly):
    return ffi_poly.intersects_poly(polygon)


def _ffi_intersect_process(ffi_list: Table, polygon: SphericalPolygon, workers=None):
    """
    Vectorizing the spherical_coordinate intersects_polygon function
    """

    start_t = time.monotonic()
    with ProcessPoolExecutor(max_workers=4) as executor:
        # Map the function across the ffi_list["polygon"]

        single_intersect_poly = functools.partial(single_intersect, polygon)
        results = executor.map(single_intersect_poly, ffi_list["polygon"], chunksize=64)
    print("took:", time.monotonic() - start_t)

    return np.array(list(results))


if __name__ == "__main__":
    # monkeypatch our version in
    ccff._ffi_intersect = _ffi_intersect_process

    footprint_file = "s3://tesscut-ops-footprints/tess_ffi_footprint_cache.json"

    all_ffis = ccff._get_s3_ffis(footprint_file, as_table=True, load_polys=True)

    coordinates = "217.42893801 -62.67949189"
    coordinates = SkyCoord(coordinates, unit="deg")

    cutout_size = ccff.parse_size_input(10)
    cone_results = ccff._ra_dec_crossmatch(all_ffis, coordinates, cutout_size, ccff.TESS_ARCSEC_PER_PX)

    print(cone_results)

```